runner/query_run.go

```diff
@@ -0,0 +1,125 @@
+						run(ctx, server, collectionOpts, logger)
+					}
+					time.Sleep(1 * time.Second)
+				}
+			}
+		}(servers[idx])
+	}
+}
+
+func run(ctx context.Context, server *state.Server, collectionOpts state.CollectionOpts, logger *util.Logger) {
+	for idx, query := range server.QueryRuns {
+		if !query.FinishedAt.IsZero() {
+			continue
+		}
+		server.QueryRunsMutex.Lock()
+		server.QueryRuns[idx].StartedAt = time.Now()
+		server.QueryRunsMutex.Unlock()
+		logger.PrintVerbose("Query run %d starting: %s", query.Id, query.QueryText)
+
+		db, err := postgres.EstablishConnection(ctx, server, logger, collectionOpts, query.DatabaseName)
+		if err != nil {
+			server.QueryRunsMutex.Lock()
+			server.QueryRuns[idx].FinishedAt = time.Now()
+			server.QueryRuns[idx].Error = err.Error()
+			server.QueryRunsMutex.Unlock()
+			continue
+		}
+		defer db.Close()
+
+		err = postgres.SetStatementTimeout(ctx, db, 60*1000)
+		if err != nil {
+			server.QueryRunsMutex.Lock()
+			server.QueryRuns[idx].FinishedAt = time.Now()
+			server.QueryRuns[idx].Error = err.Error()
+			server.QueryRunsMutex.Unlock()
+			continue
+		}
+
+		pid := 0
+		err = db.QueryRow(postgres.QueryMarkerSQL + "SELECT pg_backend_pid()").Scan(&pid)
+		if err == nil {
+			server.QueryRunsMutex.Lock()
+			server.QueryRuns[idx].BackendPid = pid
+			server.QueryRunsMutex.Unlock()
+		} else {
+			server.QueryRunsMutex.Lock()
+			server.QueryRuns[idx].FinishedAt = time.Now()
+			server.QueryRuns[idx].Error = err.Error()
+			server.QueryRunsMutex.Unlock()
+			continue
+		}
+
+		// We don't include QueryMarkerSQL so query runs are reported separately in pganalyze
+		comment := fmt.Sprintf("/* pganalyze:no-alert,pganalyze-query-run:%d */ ", query.Id)
+		prefix := ""
+		result := ""
+		if query.Type == pganalyze_collector.QueryRunType_EXPLAIN {
+			prefix = "EXPLAIN (ANALYZE, VERBOSE, BUFFERS, FORMAT JSON) "
+		}
+
+		err = db.QueryRowContext(ctx, comment+prefix+query.QueryText).Scan(&result)
+
+		if query.Type == pganalyze_collector.QueryRunType_EXPLAIN {
+			// Run EXPLAIN ANALYZE a second time to get a warm cache result
+			err = db.QueryRowContext(ctx, comment+prefix+query.QueryText).Scan(&result)
+
+			// If the EXPLAIN ANALYZE timed out, capture a regular EXPLAIN instead
+			if err != nil && strings.Contains(err.Error(), "statement timeout") {
+				prefix = "EXPLAIN (VERBOSE, FORMAT JSON) "
+				err = db.QueryRowContext(ctx, comment+prefix+query.QueryText).Scan(&result)
+			}
+		}
+
+		server.QueryRunsMutex.Lock()
+		server.QueryRuns[idx].FinishedAt = time.Now()
+		server.QueryRuns[idx].Result = result
+		if err != nil {
+			server.QueryRuns[idx].Error = err.Error()
+		}
+		server.QueryRunsMutex.Unlock()
+	}
+}
+
+// Removes old query runs that have finished
+func cleanup(server *state.Server) {
+	server.QueryRunsMutex.Lock()
+	queryRuns := make([]state.QueryRun, 0)
+	for _, query := range server.QueryRuns {
+		if time.Since(query.FinishedAt) >= 10*time.Minute {
```

Comment:
Isn't this check inverted?

Reply:
It seems to be correct to me: if the query finished running 10+ minutes ago, this code removes it.

Reply:
Ah, yes it is. Good catch.